jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/LambdaHost.kt

```diff
@@ -0,0 +1,123 @@
+import software.aws.toolkits.jetbrains.services.lambda.execution.LambdaRunConfigurationType
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.LocalLambdaRunConfiguration
+import software.aws.toolkits.jetbrains.services.lambda.execution.local.LocalLambdaRunConfigurationFactory
+import software.aws.toolkits.jetbrains.services.lambda.upload.CreateLambdaFunction
+import software.aws.toolkits.jetbrains.utils.RuntimeUtil
+
+/**
+ * Lambda Host class is used for communication with ReSharper backend through protocol
+ * for all operation related to AWS Lambda.
+ */
+class LambdaHost(project: Project) : LifetimedProjectComponent(project) {
+
+    val model = project.solution.lambdaModel
+
+    init {
+        initRunLambdaHandler()
+        initDebugLambdaHandler()
+        initCreateNewLambdaHandler()
+    }
+
+    private fun initRunLambdaHandler() =
+        model.runLambda.advise(componentLifetime) { runLambdaRequest ->
+            runConfiguration(
+                methodName = runLambdaRequest.methodName,
+                handler = runLambdaRequest.handler,
+                executor = DefaultRunExecutor.getRunExecutorInstance()
+            )
+        }
+
+    private fun initDebugLambdaHandler() =
+        model.debugLambda.advise(componentLifetime) { debugLambdaRequest ->
+            runConfiguration(
+                methodName = debugLambdaRequest.methodName,
+                handler = debugLambdaRequest.handler,
+                executor = DefaultDebugExecutor.getDebugExecutorInstance()
+            )
+        }
+
+    private fun initCreateNewLambdaHandler() =
+        model.createNewLambda.advise(componentLifetime) { createLambdaRequest ->
+            val handler = createLambdaRequest.handler
+
+            val handlerResolver = DotNetLambdaHandlerResolver()
+            val fieldId = handlerResolver.getFieldIdByHandlerName(project, handler)
+            val psiElement = RiderLambdaHandlerFakePsiElement(project, handler, fieldId).navigationElement
+            val smartPsiElementPointer = SmartPointerManager.createPointer(psiElement)
+
+            val action = CreateLambdaFunction(
+                handlerName = handler,
+                elementPointer = smartPsiElementPointer,
+                lambdaHandlerResolver = handlerResolver
+            )
+
+            val contextMap = mapOf(
+                LangDataKeys.LANGUAGE.name to LanguageUtil.getRootLanguage(psiElement)
+            )
+
+            ActionUtil.invokeAction(
+                action,
+                SimpleDataContext.getSimpleContext(contextMap, SimpleDataContext.getProjectContext(project)),
+                ActionPlaces.EDITOR_GUTTER_POPUP,
+                null,
+                null
+            )
+        }
+
+    private fun runConfiguration(methodName: String, handler: String, executor: Executor) {
+        val runManager = RunManager.getInstance(project)
```

Comment:
@abrooksv, sorry, I'm not sure I got the idea. Idea run configuration producer use Idea PSI to get element from Idea context. In Rider we do not use Idea PSI. Instead, we use R# PSI (almost do not use). 

RiderLambdaHandlerFakePsiElement I use only to have a compatibility with Idea infrastructure at some rare cases when it hard to avoid. I would get rid of this class at some point.

Reply:
The thought process was more along the lines of: There is a lot of duplicated code between this and our RunConfigurationProducer, how can we share it. Thoughts?

Reply:
Make sense, but my point was that the logic to get information for run configuration differ from an Idea approach. Using the `RiderLambdaHandlerFakePsiElement` will add extra redundant call on a backend to get a handler information while we already have this information (we got it from a backend when handle context item selection logic and invoke a run configuration). I've extract a very general method to setup `credentialProvider` and `region` for a configuration as a common part.